### PR TITLE
Devise

### PR DESCRIPTION
**Pull request by hammad-Ikhlaq-7vals**
Completed Sign up, Login etc and its UIs

### CODE REVIEW

File: app/assets/javascripts/application.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+function make_domain() {
+  var x = document.getElementById("company_name_id").value;
+  document.getElementById("domain_place").innerHTML = "Your portal will be live at: https://" + x + ".workengine.com";
+  document.getElementById("domain_id").value = "https://" + x + ".workengine.com"
+}
```
Comment (hammad-Ikhlaq-7vals) [inline app/assets/javascripts/application.js:5]: no newline at end of file

File: app/views/devise/registrations/new.html.erb [hunk 2]
```diff
@@ -1,44 +1,46 @@
...
+<br />
+<p id="domain_place"></p>
+
+  <div>
+    <%= f.submit t('users.sign_up'), class: "btn btn-primary" %>
   </div>
 <% end %>
 
-<%= render "devise/shared/links" %>
+<%= render "devise/shared/links" %>
```
Comment (hammad-Ikhlaq-7vals) [inline app/views/devise/registrations/new.html.erb:46]: new line need

**Comment by mujeeb-arshad** [comment]
Resolve conflicts

File: app/javascript/packs/application.js [hunk 3]
```diff
@@ -10,7 +10,6 @@ require("turbolinks").start()
 require("@rails/activestorage").start()
 require("channels")
 
-
```
Comment (mujeeb-arshad) [inline app/javascript/packs/application.js:13]: revert

File: config/database.yml [hunk 4]
```diff
@@ -12,9 +12,9 @@
 default: &default
   adapter: mysql2
   encoding: utf8mb4
-  pool: 5 %>
+  pool: <%= ENV.fetch("RAILS_MAX_THREADS") { 5 } %>
   username: root
-  password: root
```
Comment (mujeeb-arshad) [inline config/database.yml]: add the file to git ignore

File: config/routes.rb [hunk 5]
```diff
@@ -5,4 +5,5 @@
   end
   devise_for :users
   # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
```
Comment (mujeeb-arshad) [inline config/routes.rb]: remove extra line

File: db/migrate/20200907121405_devise_create_users.rb [hunk 6]
```diff
@@ -4,11 +4,10 @@ class DeviseCreateUsers < ActiveRecord::Migration[6.0]
   def change
     create_table :users do |t|
       ## Database authenticatable
-      t.string :name,              null: false, default: ""
+      t.string :name,               null: false, default: ""
```
Comment (mujeeb-arshad) [inline db/migrate/20200907121405_devise_create_users.rb:7]: remove space

File: db/schema.rb [hunk 7]
```diff
@@ -14,21 +14,81 @@
 
   create_table "companies", options: "ENGINE=InnoDB DEFAULT CHARSET=utf8mb4", force: :cascade do |t|
     t.string "name", default: "", null: false
-    t.string "domain", default: "", null: false
```
Comment (mujeeb-arshad) [inline db/schema.rb:17]: add this file to git ignore as well

File: db/migrate/20200907133359_create_companies.rb [hunk 8]
```diff
@@ -0,0 +1,10 @@
+class CreateCompanies < ActiveRecord::Migration[6.0]
+  def change
+    create_table :companies do |t|
+      t.string :name,   null: false, default: ""
+      t.string :domain, null: true, default: ""
```
Comment (mujeeb-arshad) [inline db/migrate/20200907133359_create_companies.rb]: can domain be ever null for a company?

File: db/migrate/20200908055721_add_company_id_to_users.rb [hunk 9]
```diff
@@ -1,5 +1,5 @@
 class AddCompanyIdToUsers < ActiveRecord::Migration[6.0]
   def change
-    add_column :users, :company_id, :integer
+    add_column :users, :company_id, :integer, null: true
```
Comment (mujeeb-arshad) [inline db/migrate/20200908055721_add_company_id_to_users.rb]: can company_id be null for a company?

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -4,6 +4,6 @@ class ApplicationController < ActionController::Base
   protected
 
   def configure_permitted_parameters
-    devise_parameter_sanitizer.permit(:sign_up, keys: [:name, :email, :password, :role, :is_confirmed, :company_id])
+    devise_parameter_sanitizer.permit(:sign_up, keys: [:name, company_attributes: [:name, :domain]])
```
Comment (muhammad-basil) [inline app/controllers/application_controller.rb]: change domain to subdomain and incorporate where change required

File: app/models/company.rb [hunk 11]
```diff
@@ -2,4 +2,7 @@ class Company < ApplicationRecord
   has_many :users
   has_many :projects
   has_many :teams
+
+  validates :domain, format: { with: /\A[a-zA-Z0-9]+\z/,
```
Comment (muhammad-basil) [inline app/models/company.rb]: subdomain

File: app/views/devise/registrations/new.html.erb [hunk 12]
```diff
@@ -1,43 +1,45 @@
...
+<h1 class="h4 text-gray-900 mb-4">Create an Account!</h1>
 
 <%= form_for(resource, as: resource_name, url: registration_path(resource_name)) do |f| %>
   <%= render "devise/shared/error_messages", resource: resource %>
 
-  <div class="field">
+  <div class="form-group">
     <%= f.label :name %><br />
-    <%= f.text_field :name%>
+    <%= f.text_field :name, class: "form-control",placeholder: "Enter Name Here", required: true %>
```
Comment (muhammad-basil) [inline app/views/devise/registrations/new.html.erb]: spacing issue on line 8 and 13

File: app/views/devise/registrations/new.html.erb [hunk 13]
```diff
@@ -1,43 +1,45 @@
...
-    <%= f.password_field :password_confirmation, autocomplete: "Password" %>
+    <%= f.password_field :password_confirmation, autocomplete: "Password", class: "form-control" , required: true %>
   </div>
 
-  <div class="field">
-    <%= f.label :role %><br />
-    <%= f.text_field :role%>
-  </div>
+  <div class="form-group" >
+    <%= f.fields_for :company, Company.new do |o| %>
```
Comment (muhammad-basil) [inline app/views/devise/registrations/new.html.erb]: what is o?
Use proper name

File: app/views/devise/registrations/new.html.erb [hunk 14]
```diff
@@ -1,43 +1,45 @@
...
+      <%= o.label :company_name %><br />
+
+      <span onchange ="make_domain()"><%= o.text_field :name, id: "company_name_id", class: "form-control" %></span><%= o.hidden_field :domain, id: "domain_id" , required: true %>
```
Comment (muhammad-basil) [inline app/views/devise/registrations/new.html.erb]: move onchange event to JS file

File: app/views/layouts/application.html.erb [hunk 15]
```diff
@@ -7,6 +7,7 @@
 
     <%= stylesheet_link_tag 'application', media: 'all', 'data-turbolinks-track': 'reload' %>
     <%= javascript_pack_tag 'application', 'data-turbolinks-track': 'reload' %>
+    <%= javascript_include_tag 'application'%>
```
Comment (muhammad-basil) [inline app/views/layouts/application.html.erb]: space

File: config/environments/development.rb [hunk 16]
```diff
@@ -34,13 +34,13 @@
...
   config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
 
+  config.action_mailer.delivery_method = :smtp
+
   # Don't care if the mailer can't send.
   config.action_mailer.raise_delivery_errors = false
 
   config.action_mailer.perform_caching = false
 
-  config.action_mailer.default_url_options = { host: 'localhost', port: 3000 }
```
Comment (muhammad-basil) [inline config/environments/development.rb:42]: revert

File: db/migrate/20200907121405_devise_create_users.rb [hunk 17]
```diff
@@ -4,11 +4,10 @@ class DeviseCreateUsers < ActiveRecord::Migration[6.0]
...
       t.string :email,              null: false, default: ""
-      t.string :encrypted_password,              null: false, default: ""
-      t.string :role,              null: false, default: ""
-      t.string :is_confirmed,              null: false, default: ""
+      t.string :encrypted_password, null: false, default: ""
+      t.string :role,               null: true, default: "OWNER"
```
Comment (muhammad-basil) [inline db/migrate/20200907121405_devise_create_users.rb:10]: Add another table for Roles, discuss with @mujeeb-arshad

File: db/migrate/20200907133359_create_companies.rb [hunk 18]
```diff
@@ -0,0 +1,10 @@
+class CreateCompanies < ActiveRecord::Migration[6.0]
+  def change
+    create_table :companies do |t|
+      t.string :name,   null: false, default: ""
+      t.string :domain, null: false, default: ""
+      t.timestamps
```
Comment (muhammad-basil) [inline db/migrate/20200907133359_create_companies.rb:6]: null: false, change :domain to :subdomain and remove default if not required

File: app/controllers/application_controller.rb (hunk 10 above)
Comment (hammad-Ikhlaq-7vals) [reply app/controllers/application_controller.rb]: done

File: app/models/company.rb (hunk 11 above)
Comment (hammad-Ikhlaq-7vals) [reply app/models/company.rb]: done

File: app/views/devise/registrations/new.html.erb (hunk 12 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/registrations/new.html.erb]: done

File: app/views/devise/registrations/new.html.erb (hunk 13 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/registrations/new.html.erb]: done

File: app/views/layouts/application.html.erb (hunk 15 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/layouts/application.html.erb]: done

File: config/environments/development.rb (hunk 16 above)
Comment (hammad-Ikhlaq-7vals) [reply config/environments/development.rb:42]: It was already written above so I removed it from here.

File: db/migrate/20200907133359_create_companies.rb (hunk 18 above)
Comment (hammad-Ikhlaq-7vals) [reply db/migrate/20200907133359_create_companies.rb:6]: done

File: app/assets/stylesheets/application.scss [hunk 19]
```diff
@@ -1 +1,14 @@
-@import "bootstrap/scss/bootstrap";
\ No newline at end of file
+@import "bootstrap/scss/bootstrap";
+
+#error_explanation {
```
Comment (hafsabatool7) [inline app/assets/stylesheets/application.scss]: Do not use ID based styling. Change to class.

File: app/models/company.rb [hunk 20]
```diff
@@ -2,4 +2,7 @@ class Company < ApplicationRecord
   has_many :users
   has_many :projects
   has_many :teams
+
+  validates :subdomain, format: { with: /\A[a-zA-Z0-9]+\z/,
+    message: "only allows letters and numbers" }
```
Comment (hafsabatool7) [inline app/models/company.rb]: Use internationalization

File: app/views/devise/confirmations/new.html.erb [hunk 21]
```diff
@@ -1,15 +1,15 @@
-<h2>Resend confirmation instructions</h2>
+<h1 class="h4 text-gray-900 mb-4">Resend confirmation instructions</h1>
```
Comment (hafsabatool7) [inline app/views/devise/confirmations/new.html.erb]: Internationalization for 'Resend confirmation instructions'

File: app/views/devise/confirmations/new.html.erb [hunk 22]
```diff
@@ -1,15 +1,15 @@
...
 
 <%= form_for(resource, as: resource_name, url: confirmation_path(resource_name), html: { method: :post }) do |f| %>
   <%= render "devise/shared/error_messages", resource: resource %>
 
-  <div class="field">
+  <div class="form-group">
     <%= f.label :email %><br />
-    <%= f.email_field :email, autofocus: true, autocomplete: "Email", value: (resource.pending_reconfirmation? ? resource.unconfirmed_email : resource.email) %>
+    <%= f.email_field :email, autofocus: true, autocomplete: "Email", value: (resource.pending_reconfirmation? ? resource.unconfirmed_email : resource.email), class: "form-control", required: true  %>
```
Comment (hafsabatool7) [inline app/views/devise/confirmations/new.html.erb]: Always use single quotes in ruby code when no string interpolation is being done
```suggestion
    <%= f.email_field :email, autofocus: true, autocomplete: 'Email', value: (resource.pending_reconfirmation? ? resource.unconfirmed_email : resource.email), class: 'form-control', required: true  %>
```

File: app/views/devise/passwords/edit.html.erb [hunk 23]
```diff
@@ -9,12 +9,12 @@
     <% if @minimum_password_length %>
       <em>(<%= @minimum_password_length %> characters minimum)</em><br />
     <% end %>
-    <%= f.password_field :password, autofocus: true, autocomplete: "New Password" %>
+    <%= f.password_field :password, autofocus: true, autocomplete: "New Password", required: true %>
```
Comment (hafsabatool7) [inline app/views/devise/passwords/edit.html.erb:12]: Internationalization

File: app/views/devise/registrations/edit.html.erb [hunk 24]
```diff
@@ -1,42 +1,42 @@
...
-  <div class="field">
+  <div class="form-group">
     <%= f.label :password_confirmation %><br />
-    <%= f.password_field :password_confirmation, autocomplete: "New Password" %>
+    <%= f.password_field :password_confirmation, autocomplete: "New Password", class: "form-control", required: true %>
   </div>
 
-  <div class="field">
+  <div class="form-group">
     <%= f.label :current_password %> <i>(we need your current password to confirm your changes)</i><br />
```
Comment (hafsabatool7) [inline app/views/devise/registrations/edit.html.erb]: Internationalization. Please verify throughout the code

File: app/views/devise/registrations/new.html.erb [hunk 25]
```diff
@@ -1,43 +1,46 @@
...
-  <div class="field">
-    <%= f.label :role %><br />
-    <%= f.text_field :role%>
-  </div>
+  <div class="form-group" >
+    <%= f.fields_for :company, Company.new do |nested_company| %>
+      <%= nested_company.label :company_name %><br />
+
+        <%= nested_company.text_field :name, id: "company_name_id", class: "form-control", required: true %>
+        <%= nested_company.hidden_field :subdomain, id: "subdomain_id" %>
```
Comment (hafsabatool7) [inline app/views/devise/registrations/new.html.erb]: indentation. Verify in the whole file

File: app/models/company.rb (hunk 20 above)
Comment (hammad-Ikhlaq-7vals) [reply app/models/company.rb]: done

File: app/assets/javascripts/application.js [hunk 26]
```diff
@@ -0,0 +1,8 @@
+$(document).ready(function(){
+  $("#company_name_id").change(function(){
+    var x = $("#company_name_id").val();
```
Comment (hafsabatool7) [inline app/assets/javascripts/application.js]: use `$(this)`
Use proper names for variables

File: app/views/devise/confirmations/new.html.erb (hunk 21 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/confirmations/new.html.erb]: done

File: app/views/devise/confirmations/new.html.erb (hunk 22 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/confirmations/new.html.erb]: done

File: app/views/devise/registrations/edit.html.erb (hunk 24 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/registrations/edit.html.erb]: done

File: app/views/devise/registrations/new.html.erb (hunk 25 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/registrations/new.html.erb]: done

File: app/assets/javascripts/application.js (hunk 26 above)
Comment (hammad-Ikhlaq-7vals) [reply app/assets/javascripts/application.js]: done

File: app/views/devise/mailer/confirmation_instructions.html.erb [hunk 27]
```diff
@@ -1,5 +1,5 @@
-<p>Welcome <%= @email %>!</p>
+<p>t('users.elcome') <%= @email %>!</p>
```
Comment (hafsabatool7) [inline app/views/devise/mailer/confirmation_instructions.html.erb]:
```suggestion
<p>t('users.welcome') <%= @email %>!</p>
```

File: app/assets/stylesheets/application.scss [hunk 28]
```diff
@@ -1,6 +1,6 @@
 @import "bootstrap/scss/bootstrap";
 
-#error_explanation {
+.error_explanation {
```
Comment (hafsabatool7) [inline app/assets/stylesheets/application.scss]: Use dashes for class names not underscore

File: app/views/devise/mailer/confirmation_instructions.html.erb (hunk 27 above)
Comment (hammad-Ikhlaq-7vals) [reply app/views/devise/mailer/confirmation_instructions.html.erb]: done

File: app/assets/stylesheets/application.scss (hunk 28 above)
Comment (hammad-Ikhlaq-7vals) [reply app/assets/stylesheets/application.scss]: done

**Pull request merged**